gui/elements/_layer_list.py

```diff
@@ -297,30 +310,74 @@ def _add(self, event):
         """Callback when an item is added to set its order and viewer.
         """
         layer = event.item
-        self._qt.layersList.insert(event.index, len(self), layer)
         layer._order = -len(self)
         layer.viewer = self.viewer
+        layer.events.select.connect(self.viewer._update_active_layers)
+        layer.events.deselect.connect(self.viewer._update_active_layers)
 
     def _remove(self, event):
         """Callback when an item is removed to remove its viewer
         and reset its order.
         """
         layer = event.item
-        self._qt.layersList.remove(layer)
         layer.viewer = None
         layer._order = 0
+        layer.events.select.disconnect(self.viewer._update_active_layers)
+        layer.events.deselect.disconnect(self.viewer._update_active_layers)
 
     def _reorder(self, event):
         """Callback when the list is reordered to propagate those changes
         to the node draw order.
         """
         for i in range(len(self)):
             self[i]._order = -i
-        self._qt.layersList.reorder()
         canvas = self.viewer._canvas
         canvas._draw_order.clear()
         canvas.update()
 
+    def _insert_reorder(self, index, insert):
```

Comment:
Would this not make more sense if `selected=None` were also a parameter?

Reply:
i don't like having selected as an input as the thing that is calling this function doesn't really know that information (though it could figure it out)

Reply:
Ah, I missed `self.**layers**._insert_reorder` at the call point.

On the other hand, I find it frustrating / icky that the behaviour of the function depends *strongly* on some secondary state of the object. But I don't have a good solution for this. You are right that `self.layers._insert_reorder(i, j, selected=self.layers.selected)` is also icky.

btw I might suggest the name `_move_layers`?

Reply:
agreed, not ideal, can maybe revisit in the future. I have renamed to `_move_layers` for now